### PR TITLE
First pass of drawing API

### PR DESCRIPTION
**Pull request by Dinnerbone**
This gets it to a state where most sane applications would work, but there's still a lot left unimplemented:

- Shape bounds may contain edge cases, it's "mostly correct" but won't account for caps or joins. Considering we have no way to query shape vs edge bounds right now, I'll leave that as an exercise for another day.
- I didn't implement gradient line styles as those aren't supported in ruffle yet, so I couldn't test them.
- There's some strange logic in Flash surrounding auto-closing of shapes. I haven't fully deciphered it and so this doesn't auto-close at all.
- No bitmap stuff.

I figure it's enough to get most basic uses of the drawing API up and running, and the rest can come when they're ready.

### CODE REVIEW

File: core/src/display_object/movie_clip.rs [hunk 1]
```diff
@@ -550,6 +578,132 @@ impl<'gc> MovieClip<'gc> {
...
+        }
+        if let Some((_, commands)) = &mut mc.current_fill {
+            commands.push(command);
+        }
+
+        if include_last {
+            if let Some(command) = mc
+                .current_fill
+                .as_ref()
+                .and_then(|(_, commands)| commands.last().cloned())
```
Comment (Herschel) [inline core/src/display_object/movie_clip.rs:661]: Don't think we need `cloned()` here, can match on `&DrawCommand::MoveTo` etc.

File: core/src/avm1/globals/point.rs [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let point = value_to_point(this.into(), avm, context)?;
+    let length = (point.0 * point.0 + point.1 * point.1).sqrt();
+    Ok(length.into())
+}
+
+fn normalize<'gc>(
```
Comment (Herschel) [inline core/src/avm1/globals/point.rs:241]: Some ones that are currently incorrect:

`(NaN, 0).normalize()`
`(undefined, 0).normalize(1)`
`(10, null).normalize(0)`

File: core/src/avm1/globals/point.rs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+}
+
+fn interpolate<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    _this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if args.len() < 3 {
+        return Ok(Value::Undefined.into());
```
Comment (Herschel) [inline core/src/avm1/globals/point.rs:202]: Should return a point `(NaN, NaN)`

File: core/src/avm1/globals/point.rs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    let a = value_to_point(args.get(0).unwrap().to_owned(), avm, context)?;
+    let b = value_to_point(args.get(1).unwrap().to_owned(), avm, context)?;
+    let f = args.get(2).unwrap().as_number(avm, context)?;
+    let result = (b.0 - (b.0 - a.0) * f, b.1 - (b.1 - a.1) * f);
```
Comment (Herschel) [inline core/src/avm1/globals/point.rs:208]: Could add a small comment about the weirdo misdocument-backwards behavior

File: core/src/avm1/globals/point.rs [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+}
+
+fn distance<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    _this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if args.len() < 2 {
+        return Ok(Value::Undefined.into());
```
Comment (Herschel) [inline core/src/avm1/globals/point.rs:158]: Should return `(NaN, NaN)`

File: core/src/avm1/globals/point.rs [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    let a = value_to_point(
+        args.get(0).unwrap_or(&Value::Undefined).to_owned(),
+        avm,
+        context,
+    )?;
+    let b = value_to_point(
+        args.get(1).unwrap_or(&Value::Undefined).to_owned(),
+        avm,
+        context,
+    )?;
```
Comment (Herschel) [inline core/src/avm1/globals/point.rs:170]: `distance` doesn't seem to actually work for non-Point objects like `{x: 1, y: 2}` -- uses `subtract`/`length` existing on prototype

**Comment by Herschel** [comment]
Thanks for this, it's great already

1) When a line/fill is flushed, the cursor position doesn't seem to be remembered, try:
```
lineStyle(10);
lineTo(10, 10);
lineStyle(1);
lineTo(30, 100);
lineStyle(10);
lineTo(40, 70);
```
Probably need to add an implicit MoveTo after each flush.

2) It'd be nice to move the custom shape drawing stuff out of MovieClip and into a separate module, maybe `graphics.rs`, both for organization and because it will be necessary for AS3 anyway (see https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/Graphics.html -- it can be used by both MovieClip and Shape in AVM2). MovieClip will own a `Graphics` (or maybe eventually an `Option<Box<Graphics>>` if we want to make it lazy initialize).

And these are more tedious nits that I don't consider blockers:

3) Matrix should not convert to/from `swf::Matrix` until actually needed somewhere, because the Matrix AS type is all `f64`s and does not use Twips, while the `swf::Matrix` type does (this causes incorrect values for large or NaN tx/ty, for example).

4) Various minor inaccuracies with weirdo Point values (see code notes).

(ugh, had a review typed up and accidentally cancelled it... so sorry if the comments are a little rushed!)

File: core/src/display_object/movie_clip.rs (hunk 1 above)
Comment (Dinnerbone) [reply core/src/display_object/movie_clip.rs:661]: We'd need to turn all the uses of x/y into `*x` `*y`, which I find less readable. A copy of the complete object happens one way or another, I think this way is preferable.

**Comment by Herschel** [comment]
Also, with the way it's implemented currently, there can be a one-frame-off delay before the drawing appears (see the attached SWF).
[draw.zip](https://github.com/ruffle-rs/ruffle/files/4663265/draw.zip)

File: core/src/display_object/movie_clip.rs (hunk 1 above)
Comment (Herschel) [reply core/src/display_object/movie_clip.rs]: If I match on `&DrawCommand::MoveTo { x, y }` with the `&`, it's smart enough to match the parameters by value

Comment (Dinnerbone) [reply core/src/display_object/movie_clip.rs]: Added a commit that does this (but by dereferencing the match entirely, per clippy suggestions)

**Comment by Dinnerbone** [comment]
> 
> 
> Also, with the way it's implemented currently, there can be a one-frame-off delay before the drawing appears (see the attached SWF).
> [draw.zip](https://github.com/ruffle-rs/ruffle/files/4663265/draw.zip)

That is true, unfortunately. I noticed it with the exporter. It's because render is immutable, but I can solve it with a cell until we figure out a better render API design.



>     1. When a line/fill is flushed, the cursor position doesn't seem to be remembered, try:

Fixed!



>     1. It'd be nice to move the custom shape drawing stuff out of `display_object::MovieClip` and into a separate module, maybe `graphics.rs` or `drawing.rs` or something, both for organization and because it will be necessary for AS3 anyway (see https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/Graphics.html -- it can be used by both MovieClip and Shape in AVM2). MovieClip will own a `Graphics` (or maybe eventually an `Option<Box<Graphics>>` if we want to make it lazy initialize).

I moved it into `drawing.rs`. I noticed there was a `Graphic` display object, but I was really unsure of its ultimate purpose. In Flash it doesn't look like there's any extra display objects for predefined or dynamic drawings, according to the debug dumps.



>     2\. Matrix should not convert to/from `swf::Matrix` until actually needed somewhere, because the Matrix AS type is all `f64`s and does not use Twips, while the `swf::Matrix` type does (this causes incorrect values for large or NaN tx/ty, for example).

That's fair, I didn't think about interior types. I did that so that we wouldn't have to duplicate existing functionality but I guess there's no way around it.



>     3\. Various minor inaccuracies with weirdo Point values (see code notes).

I may just rip Point out of this PR, it's not very topical and can wait until later!

File: core/src/drawing.rs [hunk 7]
```diff
@@ -195,17 +198,26 @@ impl Drawing {
...
+            bounding_box.encompass(x - radius, y - radius);
+            bounding_box.encompass(x + radius, y + radius);
         }
         DrawCommand::CurveTo { x1, y1, x2, y2 } => {
-            bounding_box.encompass(x1, y1);
-            bounding_box.encompass(x2, y2);
+            bounding_box.encompass(x1 - radius, y1 - radius);
+            bounding_box.encompass(x1 + radius, y1 + radius);
+            bounding_box.encompass(x2 - radius, y2 - radius);
+            bounding_box.encompass(x2 + radius, y2 + radius);
```
Comment (Herschel) [inline core/src/drawing.rs]: Better bounds for quadratic bezier: https://www.iquilezles.org/www/articles/bezierbbox/bezierbbox.htm

**Comment by Herschel** [comment]
Thanks! It's great!

> I noticed there was a `Graphic` display object, but I was really unsure of its ultimate purpose. In Flash it doesn't look like there's any extra display objects for predefined or dynamic drawings, according to the debug dumps.

If you mean our `Graphic` display object, this actually becomes a proper display object in AS3 (all pre-drawn art is a "Shape" instance, which may be a better name for it). If you mean the `Graphics` class in AS3, it's not an actual `DisplayObject`, just a container for the shared behavior that can be used from both `Sprite` and `Shape` display objects (they aren't in the same branch of the inheritance tree, so they each have a separate instance, and a reason for us to separate it, too).

**Pull request merged**